ensure bundle facade versioning

ycliuhw opened the pull request:
## Description of change

added facade versioning for BundleGet because we added `Device` support to this facade.

## QA steps

None

## Bug reference

None

ycliuhw commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Add facade upgrade for bundleGet](https://trello.com/c/zWjEIaFR/325-add-facade-upgrade-for-bundleget)

anastasiamac commented on apiserver/allfacades.go:
```diff
@@ -149,6 +149,9 @@ func AllFacades() *facade.Registry {
 	reg("Backups", 2, backups.NewFacadeV2)
 	reg("Block", 2, block.NewAPI)
 	reg("Bundle", 1, bundle.NewFacadeV1)
+	if featureflag.Enabled(feature.DeveloperMode) {
```
You do not need feature flag here. 
Also, you are upping the facade version and in this collection only one instance of it can exist.

anastasiamac commented on apiserver/facades/client/client/bundles.go:
```diff
@@ -15,12 +15,14 @@ import (
 		return params.BundleChangesResults{}, err
 	}
-	return bundleAPI.GetChanges(args)
+	apiV1 := bundle.APIv1{&bundle.APIv2{bundleAPI}}
```
That's not right. This call should not have changed. Instead your bundle.NewBundleAPI should return v2.

ycliuhw replied:
The `NewBundleAPI` returns `BundleAPI`, but facade constructors return `APIv*`

anastasiamac replied:
Right, I see. So this functionality exists on both Client and Bundle facade. And the one on the Client facade should behave as previously whereas the on Bundle facade has been revved up...

So this call should probably be bundle.NewFacadeV1().

anastasiamac replied:
By the way, if here is the only palce that you need the data returned to be different (say with no devices), then why can't we just change the output here once we get the return from bundleAPI.GetChanges()? Would it not also avoid you making a PR against bundelchanges repo?

anastasiamac commented on apiserver/facades/client/bundle/bundle.go:
```diff
@@ -103,6 +103,62 @@ func (b *BundleAPI) checkCanRead() error {
+// data. The changes are sorted by requirements, so that they can be applied in
+// order.
+// V1 GetChanges did not support device.
+func (b *APIv1) GetChanges(args params.BundleChangesParams) (params.BundleChangesResults, error) {
```
so is this pointer to a APIv1 or APIv2?
It feels very similar to what we have done in modelmanager...
For example see https://github.com/juju/juju/blob/develop/apiserver/facades/client/modelmanager/modelmanager.go#L835 vs https://github.com/juju/juju/blob/develop/apiserver/facades/client/modelmanager/modelmanager.go#L854.
This is an example of a method change in between different versions. 
You will need to keep an APIv1 call to return data without device support and have another call sitting on APIv2 providing the support. Internally they may call the same func but present data differently.

ycliuhw replied:
yes, agreed. I did think this. But here the difference between two versions are neither params nor returns. It's in the middle `GUIArgs()`. The idea solution would do whole versioning in bundlechanges repo. I was thinking to do it like that. But I am not very sure if it's worth to put this effect.

anastasiamac replied:
Really? It looks like a brand new method to me?...

anastasiamac commented on apiserver/facades/client/bundle/bundle.go:
```diff
@@ -160,6 +216,6 @@ func (b *BundleAPI) ExportBundle() (params.StringResult, error) {
-// Mask the new method from V1 API.
 // ExportBundle is not in V1 API.
-func (u *APIv1) ExportBundle() (_, _ struct{}) { return }
+// Mask the new method from V1 API.
```
I am not sure how this got in... We do not generally implement versions this way but rather using interfaces as in https://github.com/juju/juju/blob/develop/apiserver/facades/client/modelmanager/modelmanager.go#L44

ycliuhw replied:
This is existing code from develop. I just changed the doc to make linter happy

anastasiamac replied:
This looks wrong anyway. The version bump need to be changed to avoid doing this kind of placeholder, I think.

But even if you choose not to undertake this here, the comment should be starting with the method name... Maybe remove this change?

anastasiamac commented on apiserver/facades/client/bundle/bundle.go:
```diff
@@ -103,6 +103,62 @@ func (b *BundleAPI) checkCanRead() error {
+		return err
+	}
+	if err := data.Verify(verifyConstraints, verifyStorage, nil); err != nil {
+		if err, ok := err.(*charm.VerificationError); ok {
```
I hope that different scopes will take effect here. However, from readability perspective, it seems to be difficult to differentiate between which 'err'...

Also could we please add tests that exercise this code path?

ycliuhw commented:
hi @anastasiamac as you suggested, I renamed the GUIArgs to GUIArgsWithDevice in bundlechanges repo and refactored this PR for it as well.
Please take a look when you have time.
Thanks

anastasiamac commented on apiserver/facades/client/bundle/bundle.go:
```diff
@@ -143,16 +181,47 @@ func (b *BundleAPI) GetChanges(args params.BundleChangesParams) (params.BundleCh
+		results.Changes = make([]*params.BundleChange, len(changes))
+		for i, c := range changes {
+			var guiArgs []interface{}
+			switch c := c.(type) {
```
I wonder if it is needed now... I mean we are bringing in the version of bundlechanges that will have the new GUIArgsWith... so we can always call it...

Could you please explain the rationale for the 2 code paths here?

ycliuhw replied:
Because here in the `[] change`, the change is not always needs to run `GUIArgsWithDevices` but only for `ApplicationChange`.

anastasiamac commented on apiserver/facades/client/bundle/bundle.go:
```diff
@@ -162,4 +231,4 @@ func (b *BundleAPI) ExportBundle() (params.StringResult, error) {
 // Mask the new method from V1 API.
 // ExportBundle is not in V1 API.
-func (u *APIv1) ExportBundle() (_, _ struct{}) { return }
+func (b *APIv1) ExportBundle() (_, _ struct{}) { return }
```
Is this really needed? :)

ycliuhw commented:
$$merge$$

ycliuhw commented:
!!build!!

ycliuhw commented:
$$merge$$
